airflow/models/dagrun.py

```diff
@@ -561,6 +561,45 @@ def fetch_task_instances(
             tis = tis.where(TI.task_id.in_(task_ids))
         return session.scalars(tis).all()
 
+    @internal_api_call
+    def _check_last_n_dagruns_failed(self, dag_id, max_consecutive_failed_dag_runs, session):
+        """Check if last N dags failed."""
+        dag_runs = (
+            session.query(DagRun)
+            .filter(DagRun.dag_id == dag_id)
+            .order_by(DagRun.execution_date.desc())
+            .limit(max_consecutive_failed_dag_runs)
+            .all()
+        )
+
+        """ Marking dag as paused, if needed"""
+        to_be_paused = len(dag_runs) >= max_consecutive_failed_dag_runs and all(
+            dag_run.state == DagRunState.FAILED for dag_run in dag_runs
+        )
```

Comment:
```suggestion
        dag_run_states = session.scalars(
            select(DagRun.state)
            .where(DagRun.dag_id == dag_id)
            .order_by(DagRun.execution_date.desc())
            .limit(max_consecutive_failed_dag_runs)
        ).all()

        to_be_paused = len(dag_runs) >= max_consecutive_failed_dag_runs and all(
            state == DagRunState.FAILED for state in dag_run_states
        )
```

It’s likely possible to not even do the `.all()` part, but let’s keep this simple.